syntax = "proto3";

package gladys.memory;

// Memory Storage Service - Python backend
// Handles persistent storage, embeddings, and complex queries
service MemoryStorage {
    // Store a new episodic event
    rpc StoreEvent(StoreEventRequest) returns (StoreEventResponse);

    // Query events by time range
    rpc QueryByTime(QueryByTimeRequest) returns (QueryEventsResponse);

    // Query events by embedding similarity
    rpc QueryBySimilarity(QueryBySimilarityRequest) returns (QueryEventsResponse);

    // Generate embedding for text
    rpc GenerateEmbedding(GenerateEmbeddingRequest) returns (GenerateEmbeddingResponse);

    // Store/update a heuristic
    rpc StoreHeuristic(StoreHeuristicRequest) returns (StoreHeuristicResponse);

    // Query heuristics by condition match (embedding similarity)
    rpc QueryHeuristics(QueryHeuristicsRequest) returns (QueryHeuristicsResponse);

    // Query heuristics by text search (PostgreSQL tsvector)
    // Used by Rust fast path on cache miss - faster than embedding similarity
    rpc QueryMatchingHeuristics(QueryMatchingHeuristicsRequest) returns (QueryHeuristicsResponse);

    // Update heuristic confidence based on feedback (TD learning)
    rpc UpdateHeuristicConfidence(UpdateHeuristicConfidenceRequest) returns (UpdateHeuristicConfidenceResponse);

    // --- Semantic Memory (Entities & Relationships) ---

    // Store or update an entity
    rpc StoreEntity(StoreEntityRequest) returns (StoreEntityResponse);

    // Query entities by name, type, or embedding
    rpc QueryEntities(QueryEntitiesRequest) returns (QueryEntitiesResponse);

    // Store a relationship between entities
    rpc StoreRelationship(StoreRelationshipRequest) returns (StoreRelationshipResponse);

    // Get relationships for an entity (1-hop context)
    rpc GetRelationships(GetRelationshipsRequest) returns (GetRelationshipsResponse);

    // Expand context: get entity + relationships + related entities (for LLM prompts)
    rpc ExpandContext(ExpandContextRequest) returns (ExpandContextResponse);
}

// Salience Gateway Service - the "amygdala"
// Co-located with Memory per ADR-0001 ยง5.1
// Provides fast salience evaluation for incoming events
service SalienceGateway {
    // Evaluate salience of an event
    // Uses heuristics from Memory + novelty detection from Rust fast path
    rpc EvaluateSalience(EvaluateSalienceRequest) returns (EvaluateSalienceResponse);
}

// --- Events ---

message EpisodicEvent {
    string id = 1;                      // UUID
    int64 timestamp_ms = 2;             // Unix timestamp milliseconds
    string source = 3;                  // Sensor ID
    string raw_text = 4;                // Natural language description
    bytes embedding = 5;                // 384-dim float32 vector (serialized)
    SalienceVector salience = 6;
    string structured_json = 7;         // Domain-specific data as JSON string
    repeated string entity_ids = 8;     // Referenced entity UUIDs

    // Prediction instrumentation (Instrument Now, Analyze Later - ยง27)
    // Populated when LLM reasoning was triggered for this event
    float predicted_success = 9;        // LLM's prediction of action success (0.0-1.0)
    float prediction_confidence = 10;   // LLM's confidence in that prediction (0.0-1.0)
    string response_id = 11;            // Links to executive response/reasoning trace
}

message SalienceVector {
    float threat = 1;
    float opportunity = 2;
    float humor = 3;
    float novelty = 4;
    float goal_relevance = 5;
    float social = 6;
    float emotional = 7;                // -1.0 to 1.0
    float actionability = 8;
    float habituation = 9;
}

message StoreEventRequest {
    EpisodicEvent event = 1;
}

message StoreEventResponse {
    bool success = 1;
    string error = 2;
}

message QueryByTimeRequest {
    int64 start_ms = 1;
    int64 end_ms = 2;
    string source_filter = 3;           // Optional: filter by source
    int32 limit = 4;
}

message QueryBySimilarityRequest {
    bytes query_embedding = 1;          // 384-dim float32 vector
    float similarity_threshold = 2;     // Minimum cosine similarity
    int64 time_filter_hours = 3;        // Optional: limit to recent events
    int32 limit = 4;
}

message QueryEventsResponse {
    repeated EpisodicEvent events = 1;
    string error = 2;
}

// --- Embeddings ---

message GenerateEmbeddingRequest {
    string text = 1;
}

message GenerateEmbeddingResponse {
    bytes embedding = 1;                // 384-dim float32 vector
    string error = 2;
}

// --- Heuristics (CBR Design per ยง22) ---

message Heuristic {
    string id = 1;                          // UUID
    string name = 2;                        // Human-readable name

    // Condition: fuzzy matching via embedding
    string condition_text = 3;              // Human-readable, used for embedding generation
    bytes condition_embedding = 4;          // 384-dim float32 vector (serialized)
    float similarity_threshold = 5;         // Min cosine similarity to match (default 0.7)

    // Effects (salience modifiers + actions)
    string effects_json = 6;                // JSONB equivalent - action to take

    // Learning
    float confidence = 7;                   // 0.0-1.0, default 0.5
    float learning_rate = 8;                // TD learning rate, default 0.1

    // Provenance
    string origin = 9;                      // 'built_in', 'pack', 'learned', 'user'
    string origin_id = 10;                  // Pack ID, reasoning trace ID, etc.

    // Tree structure (reserved for future, not implemented in PoC)
    repeated string next_heuristic_ids = 11;
    bool is_terminal = 12;                  // Default true

    // Stats
    int64 last_fired_ms = 13;
    int32 fire_count = 14;
    int32 success_count = 15;

    // Timestamps
    int64 created_at_ms = 16;
    int64 updated_at_ms = 17;
}

message StoreHeuristicRequest {
    Heuristic heuristic = 1;
    bool generate_embedding = 2;            // If true, server generates embedding from condition_text
}

message StoreHeuristicResponse {
    bool success = 1;
    string error = 2;
    string heuristic_id = 3;                // ID of created/updated heuristic
}

message QueryHeuristicsRequest {
    // For embedding-based matching (CBR)
    string query_text = 1;                  // Text to embed and match against conditions
    bytes query_embedding = 2;              // Pre-computed embedding (if available)
    float min_similarity = 3;               // Minimum similarity threshold
    float min_confidence = 4;               // Minimum confidence filter
    int32 limit = 5;                        // Max results to return
}

// Query for text-search based matching (faster than embedding similarity)
// Uses PostgreSQL tsvector/tsquery for word-overlap matching
message QueryMatchingHeuristicsRequest {
    string event_text = 1;                  // Event text to match against heuristic conditions
    float min_confidence = 2;               // Minimum confidence filter (default 0.0)
    int32 limit = 3;                        // Max results to return (default 10)
    string source_filter = 4;               // Filter by source/domain (e.g., "minecraft", "smart_home")
}

message QueryHeuristicsResponse {
    repeated HeuristicMatch matches = 1;
    string error = 2;
}

message HeuristicMatch {
    Heuristic heuristic = 1;
    float similarity = 2;                   // Cosine similarity to query
    float score = 3;                        // similarity ร confidence
}

// --- TD Learning: Confidence Updates ---

message UpdateHeuristicConfidenceRequest {
    string heuristic_id = 1;                // UUID of the heuristic to update
    bool positive = 2;                      // True = positive feedback, False = negative
    float learning_rate = 3;                // Optional: override default 0.1
}

message UpdateHeuristicConfidenceResponse {
    bool success = 1;
    string error = 2;
    float old_confidence = 3;               // Confidence before update
    float new_confidence = 4;               // Confidence after update
    float delta = 5;                        // The change applied (for logging)
}

// --- Salience Evaluation ---

message EvaluateSalienceRequest {
    string event_id = 1;                // Event identifier
    string source = 2;                  // Sensor source ID
    string raw_text = 3;                // Event text for novelty/semantic eval
    string structured_json = 4;         // Domain-specific context as JSON
    repeated string entity_ids = 5;     // Referenced entities
    bool skip_novelty_detection = 6;    // Skip embedding-based novelty (for benchmarking)
}

message EvaluateSalienceResponse {
    SalienceVector salience = 1;        // Evaluated salience dimensions
    bool from_cache = 2;                // True if result was cached/heuristic
    string matched_heuristic_id = 3;    // ID of matched heuristic (if any)
    string error = 4;
    bool novelty_detection_skipped = 5; // True if skip_novelty_detection was set
}

// --- Semantic Memory: Entities ---

message Entity {
    string id = 1;                          // UUID
    string canonical_name = 2;              // Primary name (e.g., "Steve")
    repeated string aliases = 3;            // Alternative names (e.g., ["Steven", "Steve M."])
    string entity_type = 4;                 // person, game_character, game, place, device, etc.
    string attributes_json = 5;             // Flexible key-value storage as JSON
    bytes embedding = 6;                    // 384-dim float32 vector for similarity search
    string source = 7;                      // Where first encountered
    int64 first_seen_ms = 8;
    int64 last_seen_ms = 9;
    int32 mention_count = 10;
    int64 created_at_ms = 11;
    int64 updated_at_ms = 12;
}

message StoreEntityRequest {
    Entity entity = 1;
    bool generate_embedding = 2;            // If true, generate embedding from name + type + attributes
}

message StoreEntityResponse {
    bool success = 1;
    string error = 2;
    string entity_id = 3;                   // ID of created/updated entity
}

message QueryEntitiesRequest {
    string name_query = 1;                  // Search by name (exact or prefix)
    string entity_type = 2;                 // Filter by type
    bytes query_embedding = 3;              // Similarity search by embedding
    float min_similarity = 4;               // Minimum cosine similarity (default 0.7)
    int32 limit = 5;                        // Max results (default 10)
}

message QueryEntitiesResponse {
    repeated EntityMatch matches = 1;
    string error = 2;
}

message EntityMatch {
    Entity entity = 1;
    float similarity = 2;                   // Cosine similarity (if embedding search)
}

// --- Semantic Memory: Relationships ---

message Relationship {
    string id = 1;                          // UUID
    string subject_id = 2;                  // Entity UUID (from)
    string predicate = 3;                   // Relationship type (e.g., "has_character", "plays_in")
    string object_id = 4;                   // Entity UUID (to)
    string attributes_json = 5;             // Optional relationship properties
    float confidence = 6;                   // Confidence in this relationship (0-1)
    string source = 7;                      // Provenance: 'user', 'inferred', 'pack:minecraft'
    string source_event_id = 8;             // Optional: episodic event that established this
    int64 created_at_ms = 9;
    int64 updated_at_ms = 10;
}

message StoreRelationshipRequest {
    Relationship relationship = 1;
}

message StoreRelationshipResponse {
    bool success = 1;
    string error = 2;
    string relationship_id = 3;
}

message GetRelationshipsRequest {
    string entity_id = 1;                   // Get relationships for this entity
    string predicate_filter = 2;            // Optional: filter by predicate type
    bool include_incoming = 3;              // Include relationships where entity is object (default true)
    bool include_outgoing = 4;              // Include relationships where entity is subject (default true)
    float min_confidence = 5;               // Minimum confidence filter (default 0)
    int32 limit = 6;                        // Max results (default 50)
}

message GetRelationshipsResponse {
    repeated RelationshipWithEntity relationships = 1;
    string error = 2;
}

message RelationshipWithEntity {
    Relationship relationship = 1;
    Entity related_entity = 2;              // The entity on the "other side" of the relationship
}

// --- Context Expansion (for LLM prompts) ---

message ExpandContextRequest {
    repeated string entity_ids = 1;         // Starting entities (extracted from user query)
    int32 max_hops = 2;                     // How many hops to traverse (default 2, max 3)
    int32 max_entities = 3;                 // Max total entities to return (default 20)
    float min_confidence = 4;               // Minimum relationship confidence (default 0.5)
}

message ExpandContextResponse {
    repeated Entity entities = 1;           // All entities in the context graph
    repeated Relationship relationships = 2; // All relationships in the context
    string error = 3;
}
